File: frontend/src/lib/health/series.ts
import { toLocalDayKey, toTimeMs } from "./date";
import { filterDataByWindow, filterDataByWindowRange } from "./windows";

interface DataPoint {
  date: string;
  value: number | null;
}

export type AggregationMethod = "last" | "mean" | "max" | "sum";

export type MetricName =
  | "hrv"
  | "rhr"
  | "sleep"
  | "stress"
  | "steps"
  | "strain"
  | "calories"
  | "weight"
  | "recovery";

export const METRIC_AGGREGATION: Record<MetricName, AggregationMethod> = {
  hrv: "mean",
  rhr: "mean",
  sleep: "last",
  stress: "mean",
  steps: "last",
  strain: "max",
  calories: "last",
  weight: "last",
  recovery: "last",
};

export function toDailySeries(
  data: DataPoint[],
  method: AggregationMethod = "last",
): DataPoint[] {
  const sortedData = [...data]
    .filter((d) => d.value !== null)
    .sort((a, b) => toTimeMs(a.date) - toTimeMs(b.date));

  const dayMap = new Map<string, { values: number[]; lastTimestamp: number }>();

  for (const d of sortedData) {
    const dayKey = toLocalDayKey(d.date);
    const timestamp = toTimeMs(d.date);
    const existing = dayMap.get(dayKey) ?? { values: [], lastTimestamp: 0 };
    existing.values.push(d.value!);
    existing.lastTimestamp = Math.max(existing.lastTimestamp, timestamp);
    dayMap.set(dayKey, existing);
  }

  const result: DataPoint[] = [];
  for (const [date, { values }] of dayMap) {
    let aggregated: number;
    switch (method) {
      case "mean":
        aggregated = values.reduce((a, b) => a + b, 0) / values.length;
        break;
      case "max":
        aggregated = Math.max(...values);
        break;
      case "sum":
        aggregated = values.reduce((a, b) => a + b, 0);
        break;
      case "last":
      default:
        aggregated = values[values.length - 1];
        break;
    }
    result.push({ date, value: aggregated });
  }

  return result.sort((a, b) => toTimeMs(a.date) - toTimeMs(b.date));
}

export function toDailySeriesForMetric(
  data: DataPoint[],
  metric: MetricName,
): DataPoint[] {
  return toDailySeries(data, METRIC_AGGREGATION[metric]);
}

export function getWindowValues(
  data: DataPoint[],
  windowDays: number,
): number[] {
  const windowData = filterDataByWindow(data, windowDays).filter(
    (d) => d.value !== null,
  );
  return windowData.map((d) => d.value!);
}

export function getWindowRangeValues(
  data: DataPoint[],
  daysBack: number,
  daysBackEnd: number,
): number[] {
  const windowData = filterDataByWindowRange(
    data,
    daysBack,
    daysBackEnd,
  ).filter((d) => d.value !== null);
  return windowData.map((d) => d.value!);
}

File: src/analytics/series.py
from __future__ import annotations

from datetime import date

import pandas as pd

from .constants import METRIC_AGGREGATION
from .date_utils import (
    filter_by_window,
    filter_by_window_range,
    to_day_key,
)
from .types import DataPoint


def to_daily_series(data: list[DataPoint], method: str = "last") -> list[DataPoint]:
    rows = [
        {"date": to_day_key(d.date), "value": d.value}
        for d in data
        if d.value is not None
    ]
    if not rows:
        return []

    df = pd.DataFrame(rows)
    df["date"] = pd.to_datetime(df["date"])
    df = df.sort_values("date")

    agg_map = {"mean": "mean", "max": "max", "sum": "sum", "last": "last"}
    agg_fn = agg_map.get(method, "last")
    daily = df.groupby("date")["value"].agg(agg_fn).reset_index()

    return [
        DataPoint(date=row["date"].strftime("%Y-%m-%d"), value=float(row["value"]))
        for row in daily.to_dict("records")
    ]


def to_daily_series_for_metric(data: list[DataPoint], metric: str) -> list[DataPoint]:
    method = METRIC_AGGREGATION.get(metric, "last")
    return to_daily_series(data, method)


def get_window_values(
    data: list[DataPoint], window_days: int, ref_date: date | None = None
) -> list[float]:
    return [
        d.value
        for d in filter_by_window(data, window_days, ref_date=ref_date)
        if d.value is not None
    ]


def get_window_range_values(
    data: list[DataPoint],
    days_back: int,
    days_back_end: int,
    ref_date: date | None = None,
) -> list[float]:
    return [
        d.value
        for d in filter_by_window_range(
            data, days_back, days_back_end, ref_date=ref_date
        )
        if d.value is not None
    ]
